### ISSUE TITLE
Build MIOpen any single test independent of building MIOpen

### ISSUE THREAD

**Issue by sushkaur1** [ROCm/MIOpen#1471]
I want to build any one of the MIOpen tests (present at https://github.com/ROCmSoftwarePlatform/MIOpen/tree/amd-master/test ) without the need for building MIOpen.

Also, I tried to do the following to build a single test, but ran into include file issues:

**_cd MIOpen/test; mkdir build; cd build; cmake -DCMAKE_BUILD_TYPE=Debug .. ,  make test_conv_igemm_dynamic_xdlops_nhwc_wrw_bf16_gfx90a_**

root@ixt-sjc2-18:/root/Staging/MIOpen/test/build# **make test_conv_igemm_dynamic_xdlops_nhwc_wrw_bf16_gfx90a**
[ 50%] Building CXX object CMakeFiles/test_conv2d.dir/conv2d.o
In file included from /root/Staging/MIOpen/test/conv2d.cpp:26:
In file included from /root/Staging/MIOpen/test/conv_common.hpp:33:
/root/Staging/MIOpen/test/miopen/convolution.hpp:29:10: fatal error: 'miopen/common.hpp' file not found
#include "miopen/common.hpp"
         ^~~~~~~~~~~~~~~~~~~
1 error generated when compiling for gfx906.
CMakeFiles/test_conv2d.dir/build.make:75: recipe for target 'CMakeFiles/test_conv2d.dir/conv2d.o' failed
make[3]: *** [CMakeFiles/test_conv2d.dir/conv2d.o] Error 1
CMakeFiles/Makefile2:1392: recipe for target 'CMakeFiles/test_conv2d.dir/all' failed
make[2]: *** [CMakeFiles/test_conv2d.dir/all] Error 2
CMakeFiles/Makefile2:3401: recipe for target 'CMakeFiles/test_conv_igemm_dynamic_xdlops_nhwc_wrw_bf16_gfx90a.dir/rule' failed
make[1]: *** [CMakeFiles/test_conv_igemm_dynamic_xdlops_nhwc_wrw_bf16_gfx90a.dir/rule] Error 2
Makefile:1668: recipe for target 'test_conv_igemm_dynamic_xdlops_nhwc_wrw_bf16_gfx90a' failed
make: *** [test_conv_igemm_dynamic_xdlops_nhwc_wrw_bf16_gfx90a] Error 2

**Comment by junliume** [comment]
@pfultz2 I think this way (build test without building MIOpen) is not supported, right?

**Comment by atamazov** [comment]
@sushkaur1 If you want to build single test, do NOT `cd MIOpen/test` please. Example for `test_reduce_test`
1. configure MIOpen as usual, i.e. run `cmake ...`
2. `make -j test_reduce_test`

@junliume 
> I think this way (build test without building MIOpen) is not supported, right?

Yes, because unit tests need the subject of testing, i.e. the library itself. CMake "knows" that and builds the library if it does not exist.

**Comment by pfultz2** [comment]
> I think this way (build test without building MIOpen) is not supported, right?

No, its not supported, you need to build MIOpen with the tests. The tests use internal private classes as some are unit tests.

**Comment by sushkaur1** [comment]
I'm fine if already MIOpen debian/rpm is installed in the /opt/rocm/ path and used while building the tests.
Only I don't want to build MIOpen from source but just build the tests alone using the existing MIOpen installed.

As suggested by @atamazov , I tried doing cmake and make in test folder but facing include errors:
root@ixt-sjc2-18:/root/Staging/MIOpen/test/build# make -j test_reduce_test
[ 50%] Building CXX object CMakeFiles/test_reduce_test.dir/reduce_test.o
In file included from /root/Staging/MIOpen/test/reduce_test.cpp:27:
In file included from /root/Staging/MIOpen/test/driver.hpp:27:
/root/Staging/MIOpen/test/args.hpp:31:10: error: 'miopen/logger.hpp' file not found with <angled> include; use "quotes" instead
#include <miopen/logger.hpp>
         ^~~~~~~~~~~~~~~~~~~
         "miopen/logger.hpp"
In file included from /root/Staging/MIOpen/test/reduce_test.cpp:27:
In file included from /root/Staging/MIOpen/test/driver.hpp:27:
In file included from /root/Staging/MIOpen/test/args.hpp:31:
/root/Staging/MIOpen/test/miopen/logger.hpp:37:10: fatal error: 'miopen/each_args.hpp' file not found
#include <miopen/each_args.hpp>
         ^~~~~~~~~~~~~~~~~~~~~~
2 errors generated when compiling for gfx906.
CMakeFiles/test_reduce_test.dir/build.make:75: recipe for target 'CMakeFiles/test_reduce_test.dir/reduce_test.o' failed
make[3]: *** [CMakeFiles/test_reduce_test.dir/reduce_test.o] Error 1
CMakeFiles/Makefile2:2146: recipe for target 'CMakeFiles/test_reduce_test.dir/all' failed
make[2]: *** [CMakeFiles/test_reduce_test.dir/all] Error 2
CMakeFiles/Makefile2:2153: recipe for target 'CMakeFiles/test_reduce_test.dir/rule' failed
make[1]: *** [CMakeFiles/test_reduce_test.dir/rule] Error 2
Makefile:1044: recipe for target 'test_reduce_test' failed
make: *** [test_reduce_test] Error 2

**Comment by atamazov** [comment]
@sushkaur1 Problems with configuration phase.

> cd MIOpen/test; mkdir build; cd build; cmake -DCMAKE_BUILD_TYPE=Debug ..

When current dir is .../MIOpen/test/build, `cmake -DCMAKE_BUILD_TYPE=Debug ..` does NOT refer to the root of MIOpen, but uses .../MIOpen/test, which is wrong. You can use this:

```bash
cd MIOpen
mkdir build
cd build
cmake -DCMAKE_BUILD_TYPE=Debug ..
```
Then, run make from the same directory.

Hint: When you change cmake options, clean up build directory.

**Comment by sushkaur1** [comment]
Hi @atamazov, Thanks for your input.
But I am intentionally doing the cmake and make steps from MIOpen/test/build directory, since I am trying to use the CMakeLists.txt present at MIOpen/test directory to build the tests only and NOT the MIOpen from the source.

**Comment by atamazov** [comment]
This won't work.

**Comment by sushkaur1** [comment]
So is there any other MIOpen test/samples/examples that can be built the way I expect them to?

**Comment by atamazov** [comment]
No. But what is the final purpose? just to build the tests? or to test the existing MIOpen installation?

**Comment by sushkaur1** [comment]
The purpose is to build the MIOpen tests independent of the MIOpen building and run them using the already installed MIOpen.

**Comment by atamazov** [comment]
What you see in ./tests is _unit tests_ which can't be used in this way by definition.

But you can use MIOpenDriver in this way. Driver allows to test the most of the library's functionality.

Build the driver (`make -j MIOpenDriver`). This will build the library as well, but you can ignore it. Just run MIOpenDriver. IIRC it will use the installed library.

**Comment by sushkaur1** [comment]
But since it builds the library first, It won't work for me.
I could resolve the include errors that I reported in the above comments while using the CMakeLists.txt in test folder, but now I'm stuck with the linker errors:
[ 50%] Linking CXX executable test_conv3d
**/usr/bin/ld: cannot find -lmiopen_data**
collect2: error: ld returned 1 exit status
CMakeFiles/test_conv3d.dir/build.make:100: recipe for target 'test_conv3d' failed
make[3]: *** [test_conv3d] Error 1
CMakeFiles/Makefile2:1445: recipe for target 'CMakeFiles/test_conv3d.dir/all' failed
make[2]: *** [CMakeFiles/test_conv3d.dir/all] Error 2
CMakeFiles/Makefile2:2934: recipe for target 'CMakeFiles/test_conv_3d.dir/rule' failed
make[1]: *** [CMakeFiles/test_conv_3d.dir/rule] Error 2
Makefile:1434: recipe for target 'test_conv_3d' failed
make: *** [test_conv_3d] Error 2


Please give your inputs to resolve this?

**Comment by sushkaur1** [comment]
Any Update?

**Comment by junliume** [comment]
@sushkaur1  as mentioned by comments above, the proposed way of working is not feasible without building MIOpen first. If you need to run tests, MIOpenDriver which is installed by default should be the way to go.

**Comment by atamazov** [comment]
Indeed))

**Comment by junliume** [comment]
@pfultz2 @JehandadKhan @atamazov @DrizztDoUrden @averinevg @dmikushin 
We need to re-open this issue for more alternative proposals: **converting thousands of existing cmake tests all to gtest is not practical for any approaching release.**

CC: @sushkaur1

One way is to build test binary sets for all platforms on a non-GPU node, I think compiler supports it by "pretending" it has a certain GPU type.
Another way is to move large MIOpen dependencies (rocMLIR, CK) into installation model, so building MIOpen tests will take only a few minutes, not exactly a huge burden one cannot tolerate when testing.

**Comment by junliume** [comment]
BTW~ packing all targets into one huge binary might not a viable way, hence https://github.com/ROCmSoftwarePlatform/composable_kernel/issues/789

**Comment by JonathanLichtnerAMD** [comment]
Closing since MIOpen is being migrated to [ROCm/rocm-libraries](https://github.com/ROCm/rocm-libraries).  Please re-open there if you believe this is still an issue.
